tests/core/test_polydata.py

```diff
@@ -336,14 +337,25 @@ def test_ray_trace_origin():
 def test_multi_ray_trace(sphere):
     pytest.importorskip('rtree')
     pytest.importorskip('pyembree')
-    pytest.importorskip('trimesh')
-    origins = [[1, 0, 1], [0.5, 0, 1], [0.25, 0, 1], [0, 0, 1]]
+    trimesh = pytest.importorskip('trimesh')
+    origins = [[1, 0, 1], [0.5, 0, 1], [0.25, 0, 1], [0, 0, 5]]
     directions = [[0, 0, -1]] * 4
-    points, ind_r, ind_t = sphere.multi_ray_trace(origins, directions, retry=True)
+    points, ind_r, ind_t = sphere.multi_ray_trace(origins, directions)
     assert np.any(points)
     assert np.any(ind_r)
     assert np.any(ind_t)
 
+    # patch embree to test retry
+    with patch.object(
+        trimesh.ray.ray_pyembree.RayMeshIntersector,
+        'intersects_location',
+        return_value=[np.array([])] * 3,
+    ):
```

Comment:
I'm sorry. I am new to Python and do not understand the intent of this syntax 🤦 Could you please share a URL that will help me understand?

Reply:
https://docs.python.org/3/library/unittest.mock.html

It patches the `intersects_location` method in the Trimesh object (which it inherits from the RayMeshIntersector), so that when that method is called, instead of it actually running it just returns three empty numpy arrays. 

This is the behaviour that would occur if the ray trace ran and found no intersection points, so here it's being used to simulate the embree ray trace failing and returning nothing, such that the retry code in `multi_ray_trace` can then be tested. It also means that the `intersects_location` function doesn't actually run during the test, which makes the test a tiny bit quicker.

Reply:
Oh I see, you are mocking up the `trimesh` method in the following section. I understand now. Thanks for explaining :)
https://github.com/pyvista/pyvista/blob/2a59ac183fb7553a71faf5fd7f212b0ef115da46/pyvista/core/filters/poly_data.py#L2312-L2317